### ISSUE TITLE
OSGi integration testing

### ISSUE THREAD

**Issue by kriegfrj** [joel-costigliola/assertj-core#1888]
#### Summary

AssertJ support in OSGi is pretty good. However, due to the funky classloading tricks that it uses (particularly with assumptions and soft assertions) you can sometimes run into issues. Searching the issues for "OSGi" will reveal a number of them (some of them still open). In the osgi-test project, we have recently hit a similar issue (osgi/osgi-test#118).

This suggestion has been made before (https://github.com/joel-costigliola/assertj-core/issues/1290#issue-348166619) that we include integration tests in AssertJ that exercise the OSGi nature of AssertJ to guard against regressions and catch problems early. I think it would be nice to capture this in a dedicated issue.

#### Example

Nothing concrete yet, but in osgi-test we are going to try and achieve it by migrating the tests for our custom AssertJ assertions to leverage the Bndtools M2E integration for running integration tests inside an OSGi framework (osgi/osgi-test#121). If it works well, perhaps we consider adding this in some form to the AssertJ build.

**Comment by joel-costigliola** [comment]
It would be really great to have regression tests for OSGi, contributions welcome because we don't have the bandwith and the expertise to write this kind of tests.

**Comment by bjhargrave** [comment]
This is a difficult problem.

I started to work on this based upon how we do integration testing in https://github.com/osgi/osgi-test. To resolve the bndrun files, we put the necessary dependencies in the pom so that they are visible to the Bnd plugins as a set of possible bundles since the test execution in the OSGi runtime will need all the necessary JUnit bundles installed.

https://github.com/osgi/osgi-test/blob/0c16eefbbc79b5e8d2fee7d0fe4f122af0c93993/org.osgi.test.assertj.framework/pom.xml#L116-L120

To run JUnit Platform in an OSGi runtime, we need to use the [Bnd JUnit Platform Tester](https://github.com/bndtools/bnd/tree/master/biz.aQute.tester.junit-platform) which defines a [JUnit test engine for OSGi](https://github.com/bndtools/bnd/blob/master/biz.aQute.tester.junit-platform/src/aQute/tester/bundle/engine/BundleEngine.java). However, when the `biz.aQute.tester.junit-platform` dependency is visible in the `test` scope to the maven-surefire-plugin, it will include the BundleEngine in the "plain" JUnit execution and the BundleEngine will fail because it is being using in a non-OSGi environment.

```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.005 s <<< FAILURE! - in Bnd JUnit Platform Bundle Engine
[ERROR] Initialization Error  Time elapsed: 0.004 s  <<< ERROR!
org.junit.platform.commons.JUnitException: BundleEngine must run inside an OSGi framework
	at aQute.tester.bundle.engine.BundleEngine.discover(BundleEngine.java:59)
	at org.junit.platform.launcher.core.DefaultLauncher.discoverEngineRoot(DefaultLauncher.java:181)
	at org.junit.platform.launcher.core.DefaultLauncher.discoverRoot(DefaultLauncher.java:168)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:132)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:150)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:124)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
```

So we have a problem here where we need `biz.aQute.tester.junit-platform` as dependency for the OSGi test resolution and execution and we also cannot have `biz.aQute.tester.junit-platform` as a test scoped dependency for plain JUnit execution.

Some other ideas are needed here. We may need to refactor the repo into multiple projects where one project is the current single project (building and plain JUnit testing assertj-core) and a peer project which does the OSGi integration testing. Then each project can configure the dependencies as needed.

For now I am going to put this work item down. I just wanted to record my progress so far.

**Comment by kriegfrj** [comment]
I believe there is a way to configure JUnit Platform (in a non OSGi environment) to turn off engine autodiscovery and explicitly specify which engine you want to run. That would be the easiest  fix to this problem.

**Comment by kriegfrj** [comment]
Actually, i don't think it's necessary to have biz.aQute.tester.junit- platform as a test dependency - the embedded repo supplies it. In osgi test we simply specified junit-platform-launcher and junit-jupiter-engine in -runrequires and it works.

**Comment by bjhargrave** [comment]
> Actually, i don't think it's necessary to have biz.aQute.tester.junit- platform as a test dependency - the embedded repo supplies it.

Excellent observation! With this I was able to run OSGi integration tests. PR to come.

**Comment by bjhargrave** [comment]
> I believe there is a way to configure JUnit Platform (in a non OSGi environment) to turn off engine autodiscovery and explicitly specify which engine you want to run.

I do not think this is true. I did not see anything in the platform configuration docs about this. https://junit.org/junit5/docs/current/user-guide/#running-tests-config-params

**Comment by kriegfrj** [comment]
> I do not think this is true. I did not see anything in the platform configuration docs about this. https://junit.org/junit5/docs/current/user-guide/#running-tests-config-params

This point is now moot to an extent, as we have our other workaround, but for posterity:

Within JUnit platform's launcher API, there are explicit mechanisms for filtering test engines from the launcher request:

https://github.com/junit-team/junit5/blob/68bb2e3124da100ff00608c44fcd75dd3061f593/junit-platform-launcher/src/main/java/org/junit/platform/launcher/LauncherDiscoveryRequest.java#L61-L70

...and for turning off autoregistration altogether:

https://github.com/junit-team/junit5/blob/68bb2e3124da100ff00608c44fcd75dd3061f593/junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherConfig.java#L156-L168

These are the mechanisms I had in mind originally. However, it seems that you are right in that these APIs are not exposed by the surefire plugin, so there's no way to configure them from surefire.

But I dug a little further and realised that Surefire provides an alternative mechanism to achieve the same result: see ([Select engine and use multiple APIs](https://maven.apache.org/surefire/maven-surefire-plugin/examples/junit-platform.html#Select_engine_and_use_multiple_APIs).)

In summary, the Surefire solution is to not include the test engines as project test dependencies, but as dependencies of the maven-surefire-plugin. This usually works because your test code typically ought not to link directly to the test engine implementation(s), so there's no need for them to be on the test compile path at all - they only need to be on the classpath when the tests are running. You can then be selective about which engines you're going to include (which in turn makes surefire selective about which tests it will run).
